File: pkg/limiter/lua/leaky_bucket.lua
local key = KEYS[1]
local limit = tonumber(ARGV[1])
local interval = tonumber(ARGV[2])
local capacity = tonumber(ARGV[3])

-- {{{ Get now

local redis_time = redis.call("TIME")

local seconds = tonumber(redis_time[1])
local microseconds = tonumber(redis_time[2])

local now = math.floor((seconds * 1000) + (microseconds / 1000))

-- }}}

local rateInfo = redis.call("HMGET", key, "lastLeakTime", "water")
local lastLeakTime = tonumber(rateInfo[1])
local water = tonumber(rateInfo[2])

if not lastLeakTime then
	water = 1
	lastLeakTime = now
	redis.call("HMSET", key, "lastLeakTime", lastLeakTime, "water", water)
	return 0
end

-- {{{ leak

local timePassed = now - lastLeakTime
local intervalPassed = math.floor(timePassed / interval)
local shouldLeak = intervalPassed * limit

water = water - shouldLeak

if water < 0 then
	water = 0
end

lastLeakTime = now

-- }}}

water = water + 1

if water > capacity then
	water = capacity
	redis.call("HMSET", key, "lastLeakTime", lastLeakTime, "water", water)
	return -1
end

redis.call("HMSET", key, "lastLeakTime", lastLeakTime, "water", water)
return 0

File: pkg/limiter/leaky_bucket.go
package limiter

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// {{{ LeakyBucketLimiter

type LeakyBucketOptions struct {
	Prefix string

	Capacity int

	// to calculate rate
	Limit    int
	Interval time.Duration
}

type leakyBucketRateInfo struct {
	lastLeakTime time.Time
	water        int
}

type leakyBucketLimiter struct {
	cache    map[string]leakyBucketRateInfo
	prefix   string
	capacity int
	limit    int
	interval time.Duration

	mutex sync.Mutex
}

func NewLeakyBucketLimiter(options *LeakyBucketOptions) *leakyBucketLimiter {
	prefix := options.Prefix
	if len(prefix) == 0 {
		prefix = "rate-limit"
	}
	return &leakyBucketLimiter{
		prefix:   prefix,
		cache:    map[string]leakyBucketRateInfo{},
		capacity: options.Capacity,
		limit:    options.Limit,
		interval: options.Interval,
		mutex:    sync.Mutex{},
	}
}

func (fw *leakyBucketLimiter) leak(
	now time.Time,
	rateInfo leakyBucketRateInfo,
) leakyBucketRateInfo {
	timePassed := now.Sub(rateInfo.lastLeakTime)
	intervalsPassed := timePassed.Nanoseconds() / fw.interval.Nanoseconds()
	shouldLeak := intervalsPassed * int64(fw.limit)

	rateInfo.water -= int(shouldLeak)

	if rateInfo.water < 0 {
		rateInfo.water = 0
	}
	rateInfo.lastLeakTime = now
	return rateInfo
}

func (fw *leakyBucketLimiter) AcceptConnection(ctx context.Context, biz string) bool {
	fw.mutex.Lock()
	defer fw.mutex.Unlock()

	now := time.Now()

	key := fmt.Sprintf("%s-%s", fw.prefix, biz)
	res, ok := fw.cache[key]

	if !ok {
		// Not found
		fw.cache[key] = leakyBucketRateInfo{
			water:        1,
			lastLeakTime: now,
		}
		return true
	}

	rateInfo := fw.leak(now, res)
	rateInfo.water++

	if rateInfo.water > fw.capacity {
		rateInfo.water = fw.capacity
		fw.cache[key] = rateInfo
		return false
	}

	fw.cache[key] = rateInfo

	return true
}

// }}}
